File: 1605-minimum-number-of-days-to-make-m-bouquets/solution.java
class Solution {
    public int minDays(int[] bloomDay, int m, int k) {
        int n= bloomDay.length;
        
        //corner case handling
        
        // long val = m * k;
        // if( val > n) return -1;
        
        if(m > (n / k)) return -1;
        
        int maxi = bloomDay[0], mini = bloomDay[0];
        for(int i=0; i<n; i++){
            if( bloomDay[i] < mini) mini= bloomDay[i];
            if( bloomDay[i] > maxi) maxi = bloomDay[i];
        }
        
        int ans=0;
        int low = mini, high = maxi;
        while( low <= high){
            int mid = low + (high - low)/2;
            
            if( isPossible( mid, bloomDay, m, k)){
                ans = mid;
                high = mid-1;
            }
            else{
                low = mid +1; 
            }
        }
        return ans;
    }
    
    public boolean isPossible( int day, int []bloomDay, int m, int k){
        int noOfBouquets = 0;
        int cnt = 0;
        for( int i=0; i<bloomDay.length; i++){
            if( bloomDay[i] <= day){
                cnt++;
            }else{
                noOfBouquets+=(cnt/k);
                cnt =0;
            }
        }
        noOfBouquets+=(cnt/k);
        return noOfBouquets>=m;
    }
}

File: 1605-minimum-number-of-days-to-make-m-bouquets/solution.js
/**
 * @param {number[]} bloomDay
 * @param {number} m
 * @param {number} k
 * @return {number}
 */
var minDays = function(bloomDay, m, k) {
    if( bloomDay.length  < m*k) return -1;

    let mini = Math.min(...bloomDay);
    let maxi = Math.max(...bloomDay);

    let s=mini, e= maxi;
    let ans=-1;

    while(s<=e){
        let mid = s + Math.floor( (e-s)/2);

        if( canWeMakeMbouquets(mid, bloomDay, m, k)){
            ans = mid;
            e=mid-1;
        }else s=mid+1;
    }
    return ans;
};

function canWeMakeMbouquets(mid, bloomDay, m, k){
    let cnt=0, noOfBouquets =0;
    for(let x of bloomDay){
        if(x <= mid) cnt++;
        else{
            noOfBouquets +=  Math.floor(cnt/k);
            cnt=0;
        }
    }

    noOfBouquets +=  Math.floor(cnt/k);

    return noOfBouquets >= m ;
}
